pybamm/expression_tree/independent_variable.py

```diff
@@ -56,7 +55,7 @@ def _jac(self, variable) -> pybamm.Scalar:
         """See :meth:`pybamm.Symbol._jac()`."""
         return pybamm.Scalar(0)
 
-    def to_equation(self) -> sympy.Symbol:
+    def to_equation(self) -> "sympy.Symbol":  # noqa: F821, UP037
```

Comment:
Have done forward referencing for `sympy` here as mentioned in [PEP-484](https://peps.python.org/pep-0484/#forward-references)

Reply:
I guess my point was that we have to do forward referencing and suppressing warnings just to avoid marking a clear dependency as a dependency. I think this is a bit of a code smell and we should think about the code is organized if we really want to make sympy optional

Reply:
Usually this sort of thing is for avoiding circular imports or building abstract interfaces. I am not certain this is intended for optional dependencies:

> When a type hint contains names that have not been defined yet, that definition may be expressed as a string literal, to be resolved later.
>
> A situation where this occurs commonly is the definition of a container class, where the class being defined occurs in the signature of some of the methods.

...

> A common use for forward references is when e.g. Django models are needed in the signatures. Typically, each model is in a separate file, and has methods taking arguments whose type involves other models. Because of the way circular imports work in Python, it is often not possible to import all the needed models directly.

Reply:
Ah Sorry, guess i started this thread before reading your comment above.

> we have to do forward referencing and suppressing warnings just to avoid marking a clear dependency as a dependency. I think this is a bit of a code smell and we should think about the code is organized if we really want to make sympy optional

As this had to be done at this single specific instance I believe it doesn't affect the integrity of the code plus referring to [#3578(comment)](https://github.com/pybamm-team/PyBaMM/pull/3578#discussion_r1462422642) it should be good to come to an agreement prior to making `sympy` a required dependency.

Reply:
I agree that it should be a discussion on the sympy dependency as a whole, but I think this change goes the other way. It hides the sympy dependency even more since this is another part of the code which requires it and we removed the import.

As with my comment on the if-statement import in #3578, we are adding complexity to avoid this dependency

Reply:
> I think this change goes the other way. It hides the sympy dependency even more since this is another part of the code which requires it and we removed the import.

Yes, as this PR is meant to fix the priority issue of breaking import. Although we're not removing the import this time as it's being imported same as before #3578, we're just avoiding importing that optional dependency outside it's functional scope.

Reply:
I agree with @kratman's original comment. I would say we should add both `sympy` and `typing-extensions` as required dependencies, which should resolve the issue – the expression tree is an integral module in the PyBaMM codebase (or try to make `typing-extensions` optional rather than `sympy`). It will be a breaking change since we will have removed  the `[latexify]` extra, so we should add a note in the CHANGELOG about this as well.